src/EditorFeatures/Core/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,211 @@
+                var fireAndForget = UpdateDefinitionContext(textView, textView.Caret.Position);
+            }
+        }
+
+        void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (reason == ConnectionReason.TextViewLifetime ||
+                !textView.BufferGraph.GetTextBuffers(b => b.ContentType.IsOfType(ContentTypeNames.RoslynContentType)).Any())
+            {
+                CancellationTokenSource cancellationTokenSource;
+                if (_views.TryGetValue(textView, out cancellationTokenSource))
+                {
+                    cancellationTokenSource.Cancel();
+                    _views.Remove(textView);
+                    textView.Caret.PositionChanged -= OnTextViewCaretPositionChanged;
+                }
+            }
+        }
+
+        private void OnTextViewCaretPositionChanged(object sender, CaretPositionChangedEventArgs e)
+        {
+            var fireAndForget = UpdateDefinitionContext(e.TextView, e.NewPosition);
+        }
+
+        private async Task UpdateDefinitionContext(ITextView textView, CaretPosition caretPosition)
+        {
+            // Cancel any pending update for this view
+            _views[textView].Cancel();
+
+            // See if we moved somewhere else in a projection that we care about
+            var pointInRoslynSnapshot = caretPosition.Point.GetPoint(tb => tb.ContentType.IsOfType(ContentTypeNames.RoslynContentType), caretPosition.Affinity);
+            if (pointInRoslynSnapshot == null)
+            {
+                return;
+            }
+
+            // After a delay in case the caret moves again, find the symbol under the caret and update the context
+            var cancellationTokenSource = new CancellationTokenSource();
+            _views[textView] = cancellationTokenSource;
+            var cancellationToken = cancellationTokenSource.Token;
+
+            var foregroundTaskScheduler = TaskScheduler.FromCurrentSynchronizationContext();
+
+            var locations = await Task.Run(
+                async () => await GetContextFromPoint(pointInRoslynSnapshot, foregroundTaskScheduler, cancellationToken).ConfigureAwait(false),
+                cancellationToken).ConfigureAwait(true);
+
+            _codeDefinitionWindowService.SetContext(locations);
+        }
+
+        private async Task<ImmutableArray<CodeDefinitionWindowLocation>> GetContextFromPoint(SnapshotPoint? pointInRoslynSnapshot, TaskScheduler foregroundTaskScheduler, CancellationToken cancellationToken)
+        {
+            // TODO: Does this allocate too many tasks - should we switch to a queue like the classifier uses?
+            await Task.Delay(TimeSpan.FromMilliseconds(250), cancellationToken).ConfigureAwait(false);
+            var document = pointInRoslynSnapshot?.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
```

Comment:
We used to have some optimizations where it was desirable to call this and capture it earlier rather than doing it asynchronously, since we'd have some stuff cached that otherwise might have to be reconstructed. Is that still the case?